### ISSUE TITLE
IdsDataGrid: Wrong value being used on a row

### ISSUE THREAD

**Issue by shalini-patro12** [infor-design/enterprise-wc#2253]
**Steps To Reproduce**

Value being replaced with wrong value

**Steps To Reproduce**

1. open https://main.wc.design.infor.com/ids-data-grid/editable.html
2. serve the enterprise-wc locally, after that comment the already loaded json and keep the data as empty below i have pasted my code snippet for your reference.   "Replace your editable.ts by below editable.ts"
3. 
```
import type IdsDataGrid from '../ids-data-grid';
import type IdsPopupMenu from '../../ids-popup-menu/ids-popup-menu';
import type IdsMenuItem from '../../ids-menu/ids-menu-item';
import '../ids-data-grid';
import type { IdsDataGridColumn } from '../ids-data-grid-column';
import booksJSON from '../../../assets/data/books.json';

// Example for populating the DataGrid
const dataGrid = document.querySelector<IdsDataGrid>('#data-grid-editable')!;
const rowHeightMenu = document.querySelector<IdsPopupMenu>('#row-height-menu')!;

// Change row height with popup menu
rowHeightMenu?.addEventListener('selected', (e: Event) => {
  dataGrid.rowHeight = (e.target as IdsMenuItem).value as string;
});

(async function init() {
  // Do an ajax request
  const url: any = booksJSON;
  const columns: IdsDataGridColumn[] = [];

  // Set up columns
  columns.push({
    id: 'selectionCheckbox',
    name: 'selection',
    sortable: false,
    resizable: false,
    formatter: dataGrid.formatters.selectionCheckbox,
    align: 'center'
  });
  columns.push({
    id: 'description',
    name: 'Description',
    field: 'description',
    sortable: true,
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.text,
    editor: {
      type: 'input',
      editorSettings: {
        autoselect: true,
        dirtyTracker: true,
        validate: 'required',
      }
    },
    readonly(row: number) {
      return row % 2 === 0;
    },
    // disabled(row: number) {
    //   return row % 2 === 0;
    // },
  });
  columns.push({
    id: 'ledger',
    name: 'Ledger',
    field: 'ledger',
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.text,
    filterType: dataGrid.filters.text,
    filterOptions: {
      maxlength: 2,
      uppercase: true
    },
    uppercase: true,
    editor: {
      type: 'input',
      editorSettings: {
        autoselect: true,
        dirtyTracker: true,
        // mask: [/[a-zA-Z]/, /[a-zA-Z]/, /[a-zA-Z]/, /[a-zA-Z]/],
        maxlength: 2,
        uppercase: true
      }
    },
  });
  columns.push({
    id: 'publishDate',
    name: 'Pub. Date',
    field: 'publishDate',
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.date,
    editor: {
      type: 'datepicker',
      editorSettings: {
        validate: 'required',
        dirtyTracker: true
      }
    }
  });
  columns.push({
    id: 'publishTime',
    name: 'Pub. Time',
    field: 'publishDate',
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.time,
    editor: {
      type: 'timepicker',
      editorSettings: {
        dirtyTracker: true,
        format: dataGrid.localeAPI?.calendar().timeFormat
      }
    }
  });
  columns.push({
    id: 'price',
    name: 'Price',
    field: 'price',
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.decimal,
    formatOptions: { locale: 'en-US' },
    editor: {
      type: 'input',
      editorSettings: {
        autoselect: true,
        dirtyTracker: true,
        mask: 'number',
        maskOptions: {
          allowDecimal: true,
          integerLimit: 3,
          decimalLimit: 2
        }
      }
    },
  });
  columns.push({
    id: 'bookCurrency',
    name: 'Currency',
    field: 'bookCurrency',
    resizable: true,
    reorderable: true,
    formatter: dataGrid.formatters.dropdown,
    editor: {
      type: 'dropdown',
      editorSettings: {
        dirtyTracker: true,
        validate: 'required',
        options: [
          {
            id: '',
            label: '',
            value: ''
          },
          {
            id: 'usd',
            label: 'USD',
            value: 'usd'
          },
          {
            id: 'eur',
            label: 'EUR',
            value: 'eur'
          },
          {
            id: 'yen',
            label: 'YEN',
            value: 'yen'
          }
        ]
      }
    }
  });
  columns.push({
    id: 'transactionCurrency',
    name: 'Transaction Currency',
    field: 'transactionCurrency',
    formatter: dataGrid.formatters.text
  });
  columns.push({
    id: 'integer',
    name: 'Price (Int)',
    field: 'price',
    formatter: dataGrid.formatters.integer,
    formatOptions: { locale: 'en-US' }, // Data Values are in en-US
    editor: {
      type: 'input',
      editorSettings: {
        autoselect: true,
        dirtyTracker: true,
        mask: 'number',
        maskOptions: {
          allowDecimal: false,
          integerLimit: 3
        },
        validate: 'required'
      }
    },
  });
  columns.push({
    id: 'location',
    name: 'Location',
    field: 'location',
    formatter: dataGrid.formatters.hyperlink,
    href: '#'
  });
  columns.push({
    id: 'postHistory',
    name: 'Post History',
    field: 'postHistory',
    formatter: dataGrid.formatters.checkbox,
    align: 'center',
    editor: {
      type: 'checkbox',
      editorSettings: {
        dirtyTracker: true,
      }
    },
  });

  dataGrid.columns = columns;
  // dataGrid.idColumn = 'book';
  dataGrid.data =[];

  // const setData = async () => {
  //   const res = await fetch(url);
  //   const data = await res.json();
  //   dataGrid.data = data;
  // };

  // await setData();

  // Event Handlers
  dataGrid.addEventListener('beforecelledit', (e: Event) => {
    // Can be vetoed (<CustomEvent>e).detail.response(false);
    console.info(`Edit Started`, (<CustomEvent>e).detail);
  });

  dataGrid.addEventListener('celledit', (e: Event) => {
    console.info(`Currently Editing`, (<CustomEvent>e).detail);
  });

  dataGrid.addEventListener('endcelledit', (e: Event) => {
    console.info(`Edit Ended`, (<CustomEvent>e).detail);
  });

  dataGrid.addEventListener('cancelcelledit', (e: Event) => {
    console.info(`Edit Was Cancelled`, (<CustomEvent>e).detail);
  });

  // Example Buttons
  document.querySelector('#add-row')?.addEventListener('click', () => {
    const newId = dataGrid.datasource.currentData.length + 1;
    // const newRow = {
    //   book: newId,
    //   description: `${newId}`,
    //   ledger: 'CORE',
    //   postHistory: false,
    // };
    dataGrid.addRow({});
    dataGrid.setActiveCell(0, dataGrid.data.length - 1);
    dataGrid.editFirstCell();
  });

  document.querySelector('#delete-row')?.addEventListener('click', () => {
    dataGrid.selectedRows.reverse().forEach((row: any) => {
      dataGrid.removeRow(row.index);
    });
  });

  document.querySelector('#clear-row')?.addEventListener('click', () => {
    dataGrid.selectedRows.reverse().forEach((row: any) => {
      dataGrid.clearRow(row.index);
    });
  });
}());
```
4. save it and run it locally after that click on add icon and fill the ledger and pub time and again add one more editable field and give different data and then again click on add icon then all the content will replace by new one and that is the issue .

**Comment by mined14** [comment]
I have the same issue, it also happens when **deleting** a record as the first row record of datagrid will be changed similar on the last records of it.

**Comment by tmcconechy** [comment]
@clayinfor FYI need this one ASAP

**Comment by clayinfor** [comment]
Hey @tmcconechy, ticket ⁠[#2253](https://github.com/infor-design/enterprise-wc/issues/2253) was actually fixed by the changes I made here in ticket ⁠[#2232](https://github.com/infor-design/enterprise-wc/pull/2301/files#diff-cb16956c0c539ea8856bba566f9a9e32e7c95c4e9742993307239ae71680b645R421).
